pkg/ebpf/telemetry/errors_telemetry.go

```diff
@@ -43,213 +40,346 @@ var helperNames = map[int]string{
+			// the instrumentation instruction. The loader expects the source and destination register to be set to R0
+			// along with the correct opcode, for it to be recognized as a built-in call. However, the `call -1` does
+			// not satisfy this requirement. Therefore, we use a looser check relying on the constant to be `-1` to correctly
+			// identify the patch point.
+			//
+			// For reference IsBuiltinCall() => ins.OpCode.JumpOp() == Call && ins.Src == R0 && ins.Dst == R0
+			// We keep the first condition and augment it by checking the constant of the call.
+			if ins.OpCode.JumpOp() == asm.Call && ins.Constant == ebpfEntryTrampolinePatchCall && iter.Offset <= maxTrampolineOffset {
+				trampolinePatchSite = iter.Ins
+				trampolinePatchIndex = int(iter.Offset)
+
+				// do not break. We continue looping to get correct 'insCount'
+			}
+		}
+
+		// The patcher expects a trampoline call in all programs associated with this manager.
+		// No trampoline call is therefore an error condition. If this program is built without
+		// instrumentation we should never have come this far.
+		if trampolinePatchSite == nil {
+			return fmt.Errorf("no compiler instrumented patch site found for program %s", p.Name)
+		}
+
+		// if a specific program in a manager should not be instrumented,
+		// instrument patch point with a NOP (r1=r1)
+		if shouldSkip(p.Name) {
+			*trampolinePatchSite = asm.Mov.Reg(asm.R1, asm.R1).WithMetadata(trampolinePatchSite.Metadata)
 			continue
 		}
 
-		key := mapKey(h, m)
-		err := b.mapErrMap.Update(&key, z, ebpf.UpdateNoExist)
-		if err != nil && !errors.Is(err, ebpf.ErrKeyExist) {
-			return fmt.Errorf("failed to initialize telemetry struct for map %s", m.Name)
+		trampolineInstruction := asm.Instruction{
+			OpCode: asm.OpCode(asm.JumpClass).SetJumpOp(asm.Ja),
+			Offset: int16(insCount - trampolinePatchIndex - 1),
+		}.WithMetadata(trampolinePatchSite.Metadata)
+		*trampolinePatchSite = trampolineInstruction
+
+		// patch constant 'telemetry_program_id_key'
+		// This will be patched with the intended index into the helper errors array for this probe.
+		// This is done after patching the trampoline call, so that programs which should be skipped do
+		// not get this constant set. This would allow deadcode elimination to remove the telemetry block.
+		if err := patchConstant(p.Instructions, "telemetry_program_id_key", int64(bpfTelemetry.probeKeys[p.Name])); err != nil {
+			return fmt.Errorf("failed to patch constant 'telemetry_program_id_key' for program %s: %w", p.Name, err)
 		}
-		b.mapKeys[m.Name] = key
-	}
-	return nil
-}
 
-func (b *EBPFTelemetry) initializeHelperErrTelemetryMap() error {
-	if b.helperErrMap == nil {
-		return nil
-	}
+		// Append instrumentation blocks to the bytecode
+		var instrumentationBlock []*asm.Instruction
+		var instrumentationBlockCount int
+
+		bpfAsset, err := netbpf.ReadEBPFInstrumentationModule(bpfTelemetry.bpfDir, instrumentation.filename)
```

Comment:
Do you need to read and load this in the loop? Could it be outside once?

Reply:
This is also read in `setupForTelemetry`.  Does it need to be in both places?

Reply:
I have opted to build the instrumentation block once in `setupForTelemetry` and pass it as a parameter to the patcher.